server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputerTests.java

```diff
@@ -639,18 +638,45 @@ public void testDesiredBalanceShouldConvergeInABigCluster() {
                 if (primaryNodeId != null) {
                     dataPath.put(new NodeAndShard(primaryNodeId, shardId), "/data");
                     usedDiskSpace.compute(primaryNodeId, (k, v) -> v + thisShardSize);
+                    indexRoutingTableBuilder.addShard(
+                        newShardRouting(
+                            shardId,
+                            primaryNodeId,
+                            null,
+                            true,
+                            STARTED,
+                            AllocationId.newInitializing(inSyncIds.get(shard * (replicas + 1)))
+                        )
+                    );
+                } else {
+                    var lastAllocatedNodeId = randomValueOtherThan(null, () -> randomFrom(remainingNodeIds));
```

Comment:
Nit I think this can be 
```suggestion
                    var lastAllocatedNodeId = pickAndRemoveRandomValueFrom(remainingNodeIds);
```
`primaryNodeId` is `null` so it won't pick `null` again.